### DISCUSSION TITLE
Are Matrix working fine?

### DISCUSSION THREAD

**Discussion by MLobatoR** [beehive-lab/TornadoVM#213] [Q&A] [answered]
Hi everyone,

I'm just making a real small example to try and understand how your object Matrix works. I came across something weird...

Here's my example:
```
Matrix2DInt matrix = new Matrix2DInt(3, 2);
matrix.set(0, 0, 1);
matrix.set(0, 1, 2);
matrix.set(0, 2, 3);
		
matrix.set(1, 0, 4);
matrix.set(1, 1, 5);
matrix.set(1, 2, 6);
		
System.out.println(matrix.toString());
```
I don't know how but the output is wrong...
```
MatrixInt <2 x 3>
1 2 4 
4 5 6
```

Actually I do know.. I investigated a little more your implementation and if I got it right you are transforming this matrix into a StorageFormat, an array. But when trying to find out the position of each number the math seems wrong. Here is your method (I'm just copying and pasting it here):
```
public void set(int i, int j, int value) {
     storage[StorageFormats.toRowMajor(i, j, M)] = value;
}
```
Checking this method on class StoreFormat:
```
public static int toRowMajor(int i, int j, int yMax) {
     return (i * yMax) + j;
}
```
So when I'm trying to set line 0 column 2 the position in thos StorageFormat is 0 * 2+2 = 2 and when I'm trying to set line 1 column 0 1 * 2 + 0 = 2. Same index... 

If I'm seeing this the wrong way can you explain how Matrix work. How should I initialize it? And how to set values please?

Thank you in advance,
ML

**Comment by jjfumero** [comment]
It seems there is a bug in Tornado related to the compiler replacement
phase for the matrix API. We will analyze this case. Which backend are you
using?

Juan

**Reply by MLobatoR** [reply]
I'm sorry @jjfumero,

how can I check that?
I did tornado -version and I got this:
```
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (build 1.8.0_292-b09)
OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (build 25.292-b09-jvmci-21.1-b05, mixed mode)
```
Is this what you need to know?

Anyway the code I pasted in the previous post is directly from you github.

ML

**Reply by stratika** [reply]
hi @MLobatoR, if you build the latest `develop` branch and you run ```tornado --version```, it will display the installed backend. 

For example:
```
version=0.15-dev
branch=X
commit=Y

Backends installed: 
	 - opencl
```


I suggest to try rerun with the latest develop if possible, since I see you are running with JDK8 which is deprecated from TornadoVM. Also, I believe it makes sense to open an issue from this discussion so that you fill the template of the issues and for us to be able to use your filled information to reproduce the problem and provide a fix.

**Reply by jjfumero** [reply]
@MLobatoR , we got the bug in Tornado. It is now under code review. As @stratika mentioned. JDK8 is no longer supported in TornadoVM. Also, the Graal version we support is 22.2.0 for JDK 11 and 17.

**Reply by MLobatoR** [reply]
Ok, I will upgrade JDK and wait for the new release from tornado with the bug fixed. Can you tell me when you release the fix please, so I can try again?

Thank you in advance,
ML

**Reply by jjfumero** [reply]
@MLobatoR , the fix is already merged. 

Please, notice that we are transitioning to a new API in TornadoVM, so the latest commit (https://github.com/beehive-lab/TornadoVM/commit/7cbffcf76432940b4f9feed791789695f735de8b) includes a first phase of this new transition. 

Some changes:
- `TaskSchedule` -> `TaskGraph`
- `streamIn` -> `transferToDevice(mode, data, ... );`   This call is mandatory
- `streamOut` -> `transferToHost(data...)`

See an example: https://github.com/beehive-lab/TornadoVM/blob/master/examples/src/main/java/uk/ac/manchester/tornado/examples/compute/MatrixMultiplication2D.java#L78-L82

**Reply by MLobatoR** [reply]
Hi @jjfumero and @stratika,

Matrixes are working fine in this version.

Thank you,
ML

**Comment by jjfumero** [comment, marked as answer]
This has been solved. This message is just to mark the solution.
